File: src/main/java/utils/Range.java
package utils;

import java.util.Iterator;
import java.util.NoSuchElementException;

// emulate integer ranges in java
// code borrowed from https://stackoverflow.com/a/16570509
// it is just a syntactic sugar for classical for loops
public class Range implements Iterable<Integer> {

    private int limit;

    public Range(int limit) {
        this.limit = limit;
    }

    @Override
    public Iterator<Integer> iterator() {
        final int max = limit;
        return new Iterator<Integer>() {

            private int current = 0;

            @Override
            public boolean hasNext() {
                return current < max;
            }

            @Override
            public Integer next() {
                if (hasNext()) {
                    return current++;
                } else {
                    throw new NoSuchElementException("Range reached the end");
                }
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException("Can't remove values from a Range");
            }
        };
    }

    public static Range range(int max) {
        return new Range(max);
    }
}

File: src/main/kotlin/utils/Range.kt
package utils

// emulate integer ranges in java
// code borrowed from https://stackoverflow.com/a/16570509
// it is just a syntactic sugar for classical for loops
class Range(private val limit: Int) : Iterable<Int?> {
    override fun iterator(): MutableIterator<Int?> {
        val max = limit
        return object : MutableIterator<Int?> {
            private var current = 0
            override fun hasNext(): Boolean {
                return current < max
            }

            override fun next(): Int {
                return if (hasNext()) {
                    current++
                } else {
                    throw NoSuchElementException("Range reached the end")
                }
            }

            override fun remove() {
                throw UnsupportedOperationException("Can't remove values from a Range")
            }
        }
    }

    companion object {
        fun range(max: Int): Range {
            return Range(max)
        }
    }
}